<a href="https://colab.research.google.com/github/GenoaroSJr/DataScienceSmarkio/blob/main/Desempenho_ML.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [99]:
from google.colab import drive
import pandas as pd
import numpy as np
from sklearn.metrics import confusion_matrix

In [102]:
df = pd.read_csv("/content/drive/MyDrive/Smarkio/df_tratado.csv")
df_original = pd.read_excel("/content/drive/MyDrive/teste_smarkio_Lbs.xls")
df = df.drop(columns=['Unnamed: 0'])
df.head()

Pred_class  probabilidade    status  True_class
0           0       0.079892  approved         0.0
1           0       0.379377  approved        74.0
2           0       0.379377  approved        74.0
3           0       0.420930  approved        74.0
4           1       0.607437  approved         2.0

In [103]:
x = len(df_original)

Truee = df_original['True_class']
Pred = df_original['Pred_class']

for i in range(x):
  if (Truee[i] >= 0):
    ...
  else:
    Truee[i] = Pred[i]
print(df_original)

     Pred_class  probabilidade    status  True_class
0             2       0.079892  approved         0.0
1             2       0.379377  approved        74.0
2             2       0.379377  approved        74.0
3             2       0.420930  approved        74.0
4             2       0.607437  approved         2.0
..          ...            ...       ...         ...
595          74       0.432421  approved        74.0
596          82       0.590576  approved        82.0
597          92       0.915543  approved        92.0
598          96       0.334495  approved        96.0
599          99       0.373226  approved        22.0

[600 rows x 4 columns]


/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  # Remove the CWD from sys.path while we load stuff.


In [105]:
pred_class = df['Pred_class']

In [138]:
pred_class = np.array(pred_class)
pred_class = list(pred_class)
pred_class.count(0)
print("Errados:", pred_class.count(0), "Certos: ", pred_class.count(1))
print("soma: ", pred_class.count(0)+pred_class.count(1))

Errados: 181 Certos:  419
soma:  600


### Accuracy

Accuracy ou acuracia, diz quantas classes foram classificadas corretamente. Como já foi demonstrado acima, de 600 dados, 181 foram classificados como errados e 419 foram classificados como certos. 

Mesmo sendo uma métrica de fácil uso e interpretação, a acuracia não é uma métrica confiavel, pois:
  - Um valor alto não significa uma boa permformace do algorítimo.
  - Ela considera o mesmo peso para todos os erros.

No geral, essa métrica deve ser usada quando há a mesma proporção de exemplos para cada classe e quando a penalidade de acerto e erro para cada classe forem as mesmas.

Como o dataset disponibilizado tem uma boa variedade de dados (181 erros e 419 acertos), é valido o uso.

In [132]:
#Accuracy
from sklearn.metrics import accuracy_score
accuracy = accuracy_score(df_original['True_class'], df_original['Pred_class'])
accuracy *= 100
print("Accuracy:", accuracy)

Accuracy: 69.83333333333334


### Precisão

Esta métrica é definida pela razão entre a quantidade de exemplos classificados corretamente como positivos e o total de exemplos classificados como positivos. É a métrica que "responde" o questionamento "dos exemplos classificados como positivos, quantos realmente são?"

É uma métrica utilizada quando se quer identificar um determinado resultado de forma mais precisa, no caso, minizar os falses positivos.

In [131]:
from sklearn.metrics import  precision_score
precision = precision_score(df_original['True_class'], df_original['Pred_class'], average='weighted', zero_division=0)
precision *= 100
print("Precision:",precision)

Precision: 69.88175116236557


### F1 Score

O F1-Score é uma métrica que leva em consideração outras duas métrias, a precisão e a revocação. A F1-Score será alta se ambas as suas componentes também o forem. Portanto, tendo esse parâmetro alto, fica claro que o modelo é capaz de acertas suas predições como de recuperar os exemplos da classe de interesse. No geral, quanto maior o F1-Score, melhor.

Caso o F1-Score esteja baixo será é necessário avaliar a Precisão e o Recall. O valor baixo indica que há um problema na geração de falsos positivos ou negativos. 

Ela é boa quando o dataset possui classes desproporcionais, e o modelo não emite probabilidade. No geral, é uma métrica que substitui bem a Acuracia.

In [137]:
from sklearn.metrics import f1_score
f1_score = f1_score(df_original['True_class'], df_original['Pred_class'], average='weighted', zero_division=0)
f1_score *= 100
print(f1_score)

68.31017455927221
